src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerUdpBatcher.cs

```diff
@@ -56,106 +56,110 @@ public JaegerUdpBatcher(JaegerExporterOptions options)
+
+            try
+            {
+                await this.flushLock.WaitAsync().ConfigureAwait(false);
+
+                // flush if current batch size plus new span size equals or exceeds max batch size
+                if (this.batchByteSize + spanSize >= this.maxPacketSize)
+                {
+                    flushedSpanCount = await this.FlushAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    this.maxFlushIntervalTimer.Enabled = true;
+                }
+
+                // add span to batch and wait for more spans
+                this.currentBatch.Add(span);
+                this.batchByteSize += spanSize;
+            }
+            finally
+            {
+                this.flushLock.Release();
+            }
 
             return flushedSpanCount;
         }
 
         public async Task<int> FlushAsync(CancellationToken cancellationToken)
-        {
-            try
-            {
-                await this.flushLock.WaitAsync().ConfigureAwait(false);
-
-                this.maxFlushIntervalTimer.Enabled = false;
-
-                int n = this.currentBatch.Count;
-
-                if (n == 0)
-                {
-                    return 0;
-                }
-
-                try
-                {
-                    await this.SendAsync(this.Process, this.currentBatch, cancellationToken).ConfigureAwait(false);
-                }
-                finally
-                {
-                    this.currentBatch.Clear();
-                    this.batchByteSize = this.processByteSize.Value;
-                }
-
-                return n;
-            }
-            finally
-            {
-                this.flushLock.Release();
-            }
+        {
+            this.maxFlushIntervalTimer.Enabled = false;
+
+            int n = this.currentBatch.Count;
+
+            if (n == 0)
+            {
+                return 0;
+            }
+
+            try
+            {
+                await this.SendAsync(this.Process, this.currentBatch, cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                this.currentBatch.Clear();
+                this.batchByteSize = this.processByteSize.Value;
+            }
+
+            return n;
         }
 
-        public virtual Task<int> CloseAsync(CancellationToken cancellationToken) => this.FlushAsync(cancellationToken);
+        public virtual Task<int> CloseAsync(CancellationToken cancellationToken) => this.LockedFlushAsync(cancellationToken);
 
         public void Dispose()
         {
             // Do not change this code. Put cleanup code in Dispose(bool disposing).
             this.Dispose(true);
         }
 
+        protected async Task<int> LockedFlushAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                await this.flushLock.WaitAsync().ConfigureAwait(false);
```

Comment:
I believe the `WaitAsync` needs to be before the `try` block. You'd only release if you successfully acquire the lock

Reply:
Agree, I will do changes and test accordingly since there are other methods with the same pattern that I was not updating

Reply:
@luisjfdez where did you spot this pattern? It'd be best to align so either I'm wrong here, or it's best if we change the other locations too. /cc @lmolkova

Reply:
Yeah, locks need to be acquired before the try/catch/finally.